root/static/scripts/tests/i18n/expand2.js

```diff
@@ -0,0 +1,127 @@
+    'A <a href="http://www.apple.com">darn pear</a>',
+  );
+
+  expandTest(
+    'A {apple_fruit|darn {apple}}',
+    {apple_fruit: 'http://www.apple.com', apple: React.createElement('i', null, 'pear')},
+    'A <a href="http://www.apple.com">darn <i>pear</i></a>',
+  );
+
+  expandTest(
+    'A {apple_fruit|{apple}}',
+    {
+      apple_fruit: {
+        className: 'link',
+        href: 'http://www.apple.com',
+        target: '_blank',
+      },
+      apple: 'pear',
+    },
+    'A <a class="link" href="http://www.apple.com" target="_blank">pear</a>',
+  );
+
+  expandTest(
+    'A {apple_fruit|{apple}}',
+    {
+      apple_fruit: 'http://www.apple.com',
+      apple: '<pears are="yellow, green & red">',
+    },
+    'A <a href="http://www.apple.com">&lt;pears are=&quot;yellow, green &amp; red&quot;&gt;</a>',
+  );
+
+  expandTest(
+    'A {apple_fruit|^(apple|pear)[sz.]?$}',
+    {apple_fruit: 'http://www.apple.com'},
+    'A <a href="http://www.apple.com">^(apple|pear)[sz.]?$</a>',
+  );
+
+  expandTest(
+    'A {apple_fruit|<b><strong>{prefix} APPLE!</strong></b>}',
+    {apple_fruit: 'http://www.apple.com', prefix: 'dang'},
+    'A <a href="http://www.apple.com"><b><strong>dang APPLE!</strong></b></a>',
+  );
+
+  expandTest('{x:y|}', null, '{x:y|}');
+  expandTest('{x:y|}', {x: true}, 'y');
+  expandTest('{x:y|}', {x: false}, '');
+  expandTest('{x:<b>|</b>|}', {x: true}, '<b>|</b>');
+
+  expandTest('{x:|y}', null, '{x:|y}');
+  expandTest('{x:|y}', {x: true}, '');
+  expandTest('{x:|y}', {x: false}, 'y');
+  expandTest('{x:|%y%}', {x: ''}, 'y');
+
+  expandTest('{x:%|}', {x: ''}, '');
+  expandTest('{x:%|}', {x: '%'}, '%');
+  expandTest('{x:%|}', {x: '&percnt;'}, '&amp;percnt;');
+  expandTest('{x:a%c|}', {x: 'b'}, 'abc');
+  expandTest('{x:a&percnt;c|}', {x: 'b'}, 'a%c');
+
+  expandTest('<a href="{x}"></a>', {x: '&'}, '<a href="&amp;"></a>');
+  expandTest('<a href="{x:%|}"></a>', {x: '%'}, '<a href="%"></a>');
+  expandTest('<a href="{x:%|}"></a>', {x: '&percnt;'}, '<a href="&amp;percnt;"></a>');
+  expandTest('<a href="<{x:&&percnt;|}"></a>', {x: 'b'}, '<a href="&lt;&amp;%"></a>');
+
+  expandTest('{x:{y|% :)}|{z|:(}}', {x: 'hi', y: 'www'}, '<a href="www">hi :)</a>');
+  expandTest(
+    '{x:{y|% :)}|{z|<br class="{zz}" />:(}}',
+    {x: '', z: 'www', zz: '"<>"'},
+    '<a href="www"><br class="&quot;&lt;&gt;&quot;"/>:(</a>',
+  );
+  expandTest('{a:%{b:%|}%|}', {a: '0', b: ''}, '00');
+  expandTest('{a:%{b:%|}%|}', {a: '0', b: '1'}, '010');
+  expandTest(
+    '<ul>{a:<li key="0">%</li>{b:<li key="1">%</li>|}|}</ul>',
+    {a: '0', b: '1'},
+    '<ul><li>0</li><li>1</li></ul>',
+  );
+
+  expandTest(
+    '<a href="javascript:alert(\'HAx0r\')"></a>',
+    null,
+    '&lt;a href=&quot;javascript:alert(&#x27;HAx0r&#x27;)&quot;&gt;&lt;/a&gt;',
+  );
+  expandTest(
+    '<script>alert("HAx0r")</script>',
+    null,
+    '&lt;script&gt;alert(&quot;HAx0r&quot;)&lt;/script&gt;',
+  );
```

Comment:
That would be nice to capture and test thrown error for the two latest samples.

That should not prevent merging this pull request anyway.

Reply:
Some test suggestions for attributes filtering:
```html
<br size="&{alert('HAx0r')}">
<li style="behavior: url(http://hax.0r);">
<span style="background-image: url(javascript:alert('HAx0r'))">
```

Reply:
I don't know which browser is still vulnerable to the first example which comes from [OWASP](https://www.owasp.org/index.php/XSS_Filter_Evasion_Cheat_Sheet).

Looking at it again, testing for javascript protocol scheme might not be that safe. Other examples on the same page defeat the current [`jsProtocol`](https://github.com/metabrainz/musicbrainz-server/pull/668/files#diff-24e22d15e660de887ecc6765c640286cR31) pattern: `javacript&colon;`, `&NewLine;javascript:`. Moreover, some other protocol schemes may be used for XSS: `applescript`, `vbscript`…  It is difficult to assume that there is no/won't be any other way to execute code in there. It might be safer to whitelist internal paths (starting with `/`) and protocols `http` and `https`, which is all we need for now.

Reply:
Good point! I'll switch it to a whitelist instead. Also, thankfully, that bizarre `&{}` syntax seems to be a [non-standard proprietary remnant of Netscape <= 4](https://security.stackexchange.com/questions/64925/xss-technique-javascript-include/64926#64926) and not usable in any other browser.